### PR TITLE
[sival,uart] Add test for UART baud rates

### LINKED ISSUE #19868: [chip-test] chip_sw_uart_rand_baudrate
### Test point name

chip_sw_uart_rand_baudrate

### Host side component

Rust

### OpenTitanTool infrastructure implemented

Yes

### Silicon Validation (SiVal)

Yes

### Emulation Targets

- [ ] None
- [ ] CW310
- [X] Hyperdebug + CW310

### Contact person

@rswarbrick

### Checklist

Please fill out this checklist as items are completed. Link to PRs and issues as appropriate.
- [x] Device-side (C) component developed
- [x] Bazel build rules developed
- [ ] Host-side component developed
- [ ] Test added to dvsim nightly regression (and passing at time of checking)
- [ ] For SiVal test cases, test is running relevant FPGA or silicon regression

> estimate 3

### PR DESCRIPTION
**Pull request by jwnrt**
This test runs each UART with the system loopback enabled at various Baud rates and checks that the data is written and read correctly.

Resolves #19868.

Is it okay to use the system loopback here? If needed, I can add a host side to check the data but that will complicate the test.

Two of the Baud rates specified in the test plan cannot be checked on the FPGA because it doesn't clock high enough, but they should work fine on fully clocked silicon.

### CODE REVIEW

File: sw/device/tests/uart_baud_rate_test.c [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    size_t baud_count =
+        kDeviceType == kDeviceSilicon ? kBaudCountSilicon : kBaudCountFpga;
+
+    // Check every baud rate is sent and received okay.
+    for (size_t baud_idx = 0; baud_idx < baud_count; ++baud_idx) {
+      uint32_t baud_rate = kBauds[baud_idx];
+      if (uart_idx != 0) {
+        LOG_INFO("Testing UART%d at baudrate %d", uart_idx, baud_rate);
+      }
+      test_uart_baud(baud_rate);
```
Comment (engdoreis) [inline sw/device/tests/uart_baud_rate_test.c:85]: I would use the macro `EXECUTE_TEST` here.

**Review by engdoreis — CHANGES_REQUESTED** [review]
By using the internal loopback we are not really validating the silicon, there are other parts that can impact the baud which are not exercised by the loopback (i.e pinmux).

**Comment by jwnrt** [comment]
Do you think I could somehow use the pinmux to create a loopback that would exercise the test better? I'd like to avoid having to use an opentitanlib test harness if possible because of how fiddly it is.

**Comment by jwnrt** [comment]
I've changed the test to use a host side harness, but it's dependent on #21072 so will be failing for now. It passes when all the baud rates being tested are `115200` so can still be reviewed.

**Review by rswarbrick — APPROVED** [review]

File: sw/device/tests/uart_baud_rate_test.c [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  LOG_INFO("Receiving data...");
+
+  CHECK_DIF_OK(dif_uart_watermark_rx_set(&uart, kDifUartWatermarkByte1));
+
+  uint8_t data[sizeof(kSendData)] = {0};
+  for (size_t i = 0; i < sizeof(data); ++i) {
+    CHECK_DIF_OK(dif_uart_byte_receive_polled(&uart, &data[i]));
+  }
+
+  for (size_t i = 0; i < sizeof(data); ++i) {
```
Comment (pamaury) [inline sw/device/tests/uart_baud_rate_test.c:69]: I think the loop is redundant?

Comment (jwnrt) [reply sw/device/tests/uart_baud_rate_test.c:69]: Oops, yes it is

File: sw/device/tests/uart_baud_rate_test.c [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  CHECK_DIF_OK(dif_uart_configure(
+      &uart, (dif_uart_config_t){
+                 .baudrate = (uint32_t)baud_rate,
+                 .clk_freq_hz = (uint32_t)kClockFreqPeripheralHz,
+                 .parity_enable = kDifToggleDisabled,
+                 .parity = kDifUartParityEven,
+                 .tx_enable = kDifToggleEnabled,
+                 .rx_enable = kDifToggleEnabled,
+             }));
+  OTTF_WAIT_FOR(baud_rate == 0, kTestTimeoutMillis);
```
Comment (jwnrt) [inline sw/device/tests/uart_baud_rate_test.c:58]: This is a bad way of synchronising with the host and I want to replace it with something nicer

**Comment by jesultra** [comment]
> I've changed the test to use a host side harness, but it's dependent on #21072 so will be failing for now. It passes when all the baud rates being tested are `115200` so can still be reviewed.

I apologize for the delay.  PR #21072 has merged now, meaning that opentitanlib should be able to control baud rate and parity on HyperDebug UART ports.

**Comment by jwnrt** [comment]
> I apologize for the delay.

No problem, I hope I didn't rush you. Thanks for implementing that!

**Comment by jwnrt** [comment]
I cannot work out what's wrong with `_fpga_cw310_test_rom` or `_rom_with_fake_keys`. It's like no data is going over the lines at all. I had a logic analyzer hooked up to the pins on the HyperDebug but didn't see the data moving across.

Anyway, I've just removed those two targets from the list of tests. I hope that's acceptible.

**Review by engdoreis — APPROVED** [review]

**Comment by jwnrt** [comment]
Reopening until https://github.com/lowRISC/opentitan/pull/21513 updates the test plan

**Pull request merged**